pkg/components/pubsub/pluggable.go

```diff
@@ -0,0 +1,196 @@
+
+	"github.com/dapr/dapr/pkg/components"
+	"github.com/dapr/dapr/pkg/components/pluggable"
+	proto "github.com/dapr/dapr/pkg/proto/components/v1"
+
+	"github.com/dapr/kit/logger"
+
+	"github.com/pkg/errors"
+)
+
+// grpcPubSub is a implementation of a pubsub over a gRPC Protocol.
+type grpcPubSub struct {
+	*pluggable.GRPCConnector[proto.PubSubClient]
+	// features the list of state store implemented features features.
+	features []pubsub.Feature
+	logger   logger.Logger
+}
+
+// Init initializes the grpc pubsub passing out the metadata to the grpc component.
+// It also fetches and set the component features.
+func (p *grpcPubSub) Init(metadata pubsub.Metadata) error {
+	if err := p.Dial(metadata.Name); err != nil {
+		return err
+	}
+
+	protoMetadata := &proto.MetadataRequest{
+		Properties: metadata.Properties,
+	}
+
+	// TODO Static data could be retrieved in another way, a necessary discussion should start soon.
+	// we need to call the method here because features could return an error and the features interface doesn't support errors
+	featureResponse, err := p.Client.Features(p.Context, &proto.FeaturesRequest{})
+	if err != nil {
+		return err
+	}
+
+	p.features = make([]pubsub.Feature, len(featureResponse.Feature))
+	for idx, f := range featureResponse.Feature {
+		p.features[idx] = pubsub.Feature(f)
+	}
+
+	_, err = p.Client.Init(p.Context, &proto.PubSubInitRequest{
+		Metadata: protoMetadata,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// Features lists all implemented features.
+func (p *grpcPubSub) Features() []pubsub.Feature {
+	return p.features
+}
+
+// Publish publishes data to a topic.
+func (p *grpcPubSub) Publish(req *pubsub.PublishRequest) error {
+	_, err := p.Client.Publish(p.Context, &proto.PublishRequest{
+		Topic:      req.Topic,
+		PubsubName: req.PubsubName,
+		Data:       req.Data,
+		Metadata:   req.Metadata,
+	})
+	return err
+}
+
+type messageHandler = func(*proto.Message)
+
+// adaptHandler returns a non-error function that handle the message with the given handler and ack when returns.
+//
+//nolint:nosnakecase
+func (p *grpcPubSub) adaptHandler(streamingPull proto.PubSub_PullMessagesClient, handler pubsub.Handler) messageHandler {
+	ctx := streamingPull.Context()
+	return func(msg *proto.Message) {
+		m := pubsub.NewMessage{
+			Data:        msg.Data,
+			ContentType: &msg.ContentType,
+			Topic:       msg.Topic,
+			Metadata:    msg.Metadata,
+		}
+		var ackError *proto.AckMessageError
+
+		if err := handler(ctx, &m); err != nil {
+			p.logger.Errorf("error when handling message on topic %s", msg.Topic)
+			ackError = &proto.AckMessageError{
+				Message: err.Error(),
+			}
+		}
+
+		if err := streamingPull.Send(&proto.MessageAcknowledgement{
+			MessageId: msg.Id,
+			Error:     ackError,
+		}); err != nil {
+			p.logger.Errorf("error when ack'ing message %s from topic %s", msg.Id, msg.Topic)
+		}
+	}
+}
+
+const metadataSubscriptionID = "subscription-id"
```

Comment:
@artursouza @yaron2 are you guys ok with having subscription-id as a stream metadata?

I found this a better solution semantically speaking compared to send two different type of messages over same stream and require a certain order (i.e subscription message should be sent first than acknowledgement messages)